libdnf/dnf-context.c

```diff
@@ -2008,6 +2008,128 @@ dnf_context_invalidate(DnfContext *context, const gchar *message)
                                 DNF_CONTEXT_INVALIDATE_FLAG_RPMDB);
 }
 
+
+/*
+ * dnf_context_clean_file_list:
+ * @file_path_list: the files waiting to be clean
+ * @state: a #DnfState instance
+ * @error: GError for error tracking
+ *
+ * Clean files based on file paths
+ *
+ */
+static gboolean
+dnf_context_clean_file_list(gchar** file_path_list, DnfState *state, GError **error)
+{
+    /* The case where list is empty, will return */
+    if (file_path_list == NULL || !*file_path_list)
+        return TRUE;
+    guint length;
+    /* find the length of total files to delete and set it to state */
+    length = g_strv_length(file_path_list);
+
+    dnf_state_set_number_steps(state, length);
+
+    for (char **iter = file_path_list; iter && *iter; iter++) {
+        /* if the file is a directory do the recursive remove */
+        const char* file_path = *iter;
+        if (g_file_test(file_path, G_FILE_TEST_IS_DIR)) {
+            if (!dnf_remove_recursive(file_path, error))
+                return FALSE;
+        }
+        else {
+            if (!dnf_file_cleanup(file_path, error))
+                return FALSE;
+        }
+        /* update the progress */
+        if (!dnf_state_done(state, error))
+            return FALSE;
+    }
+    return TRUE;
+}
+
+/**
+ * dnf_context_clean_cache:
+ * @context: a #DnfContext instance.
+ * @flags: #DnfContextCleanFlags flag, e.g. %DNF_CONTEXT_CLEAN_EXPIRE_CACHE
+ * @error: a #GError instance
+ *
+ * Clean the cache content in the current context cache directory based
+ * on the context flags
+ *
+ * Currently support four different clean flags:
+ * 1: DNF_CONTEXT_CLEAN_EXPIRE_CACHE: Elminate the local data and outdated cache entries
+ * 2: DNF_CONTEXT_CLEAN_PACKAGES: Eliminate any cached packages
+ * 3: DNF_CONTEXT_CLEAN_METADATA: Eliminate all of the files which libdnf uses to determine remote availability of packages
+ * 4: DNF_CONTEXT_CLEAN_ALL_REPOS: Eliminate all of the content in the cache directory
+ **/
+gboolean
+dnf_context_clean_cache(DnfContext *context,
+                        GError **error,
```

Comment:
ah, I decided to move error to the second argument is because.. if I did not do that.. the format in dnf-context.h will look a bit weird because `DnfContextCleanFlags flags` is a longer term than the error, which will look like the following: 
```
gboolean         dnf_context_clean_cache                (DnfContext     *context,          
                                                         DnfContextCleanFlags      flags
                                                         GError         *error );
```









Tho, I am also fine with error argument goes last :).